websocket/connection.go

```diff
@@ -138,14 +143,16 @@ func newConnection(target string, messageChan chan []byte) *ManagedConnection {
 
 // connect tries to establish a websocket connection.
 func (c *ManagedConnection) connect() (err error) {
+	c.connectionLock.Lock()
```

Comment:
Ah, I see, we backoff here.
So I see two problems here, if the target is unavailable:
- the lock is held basically "forever"  (we wait 20 exponential steps starting with 100ms, i.e. last wait is 19seconds, and if we consider jitter then even more, ~64 seconds in total without jitter). Lock should be held only for `c.connection=conn` part, rather than wholly.
- the <-c.closeChan won't be read during all this time. Considering we continue the loop if this returns an error, the backoff here should be minimal (or simply moved above).

Reply:
Now I read the whole file in detail, I can see how this *will* work, but I think my comment still stands. While `connect()` is spinning to return `Shutdown()` cannot finish, since it waits for this block to end and it's more than a minute in the worst case (i.e. k8s might force kill the pod by then).
I think select on closeChan and return error to the wait function is an appropriate way to deal with this.

Reply:
Agreed, added a close signal to all long-running loops.

Reply:
The point of holding the lock in the outer scope though is to prevent the Send and Read to even try to do something while the connection is still being established. In theory that'd cause less dropped messages?

Reply:
I think the only interesting case that is remaining is:
- server sends messages in a stream
- this libraries client tries to shut down. The following things happen in parallel:
  - client stops consuming from `messagechan`
  - read consumes 2 messages (one is put in the channel the other is blocking `<-`)
- client calls `Shutdown`.

The race between `read()` blocked on writing to `messageChan` and `Shutdown`, causes `read()` to block forever and `Shutdown` to block forever as well, since it can't acquire the connection lock.

Probably the easiest way to solve this is to drain the channel in `Shutdown()` (we can presume library client is no longer interested in data, since Shutdown has been invoked).

Reply:
I see your point, I'd rather not do something to that channel other than writing to it since we don't "own" it. WDYT about documenting that the user must drain the channel even after calling Shutdown to avoid deadlocks?

(Receiving messages really isn't used in any proper production code so I'm trying to not put too much effort into it since I want the logging fixes etc. out there)

Reply:
Well it's two things: `go c.Shutdown()` && `for {select {case <-data:default:}}`.
I need to think if/how we can solve this in a more elegant way, but for now, specifying an example usage in a comment should suffice.